<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Day 5</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="/kogo-2018/css/main.css">
  <link rel="canonical" href="http://localhost:4000/kogo-2018/day5/">
  <link rel="alternate" type="application/rss+xml" title="KOGO HTS Workshop 2018" href="http://localhost:4000/kogo-2018/feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/kogo-2018/">KOGO HTS Workshop 2018</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        <a class="page-link" href="/kogo-2018/day1/">Day 1</a>
        <a class="page-link" href="/kogo-2018/day2/">Day 2</a>
        <a class="page-link" href="/kogo-2018/setup/">Setup</a>
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <article class="post-content">
    <h1 id="day-5--unified-analytical-group-projects">Day 5.  Unified Analytical Group Projects</h1>
<p>On the final day of this bootcamp, we will group up with the other members of our table and embark on a joint effort to take what we have learned this past week and use it to explore a large-scale biological data set in a collaborative fashion. We will make use of the <a href="http://www.geuvadis.org/web/geuvadis/RNAseq-project">Geuvadis Project</a> which combines data on genetic variation from the <a href="http://www.1000genomes.org/">1000 Genomes Project</a> with gene expression measurements derived from RNA-sequencing generated in <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3918453/">Lappalainen et al, Nature 2013</a> to detect and visualize potential expression quantative trait loci (<a href="https://en.wikipedia.org/wiki/Expression_quantitative_trait_loci">eQTL</a>).</p>

<p>Typically, such research projects can take a very long time to generate the data and analyze the results. For the purposes of this bootcamp, we will be using a small subset of these data and will attempt to recreate the published results over these regions. Our goal is to give you a taste of what types of data exploration are now available to you with the simple yet powerful biocomputing tools you have learned and to serve as a foundation for your future research endeavors.</p>

<p><br /></p>

<h3 id="schedule">Schedule:</h3>

<table>
  <thead>
    <tr>
      <th style="text-align: center">Session</th>
      <th style="text-align: center">Time</th>
      <th style="text-align: left">Topics</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: center">I</td>
      <td style="text-align: center">9:00-10:15 AM</td>
      <td style="text-align: left"><strong>Introduction to eQTLs and Overview of Project</strong></td>
    </tr>
    <tr>
      <td style="text-align: center"> </td>
      <td style="text-align: center">10:15-10:30 AM</td>
      <td style="text-align: left">Coffee Break</td>
    </tr>
    <tr>
      <td style="text-align: center">II</td>
      <td style="text-align: center">10:30-12:00 AM</td>
      <td style="text-align: left"><strong>Obtaining, Parsing and Formatting Data</strong></td>
    </tr>
    <tr>
      <td style="text-align: center"> </td>
      <td style="text-align: center">12:00-1:00 PM</td>
      <td style="text-align: left">Lunch</td>
    </tr>
    <tr>
      <td style="text-align: center">III</td>
      <td style="text-align: center">1:00-2:15 PM</td>
      <td style="text-align: left"><strong>Parallel Association Testing and Visualization</strong></td>
    </tr>
    <tr>
      <td style="text-align: center"> </td>
      <td style="text-align: center">2:15-2:30 PM</td>
      <td style="text-align: left">Coffee Break</td>
    </tr>
    <tr>
      <td style="text-align: center">IV</td>
      <td style="text-align: center">2:30-4:00 PM</td>
      <td style="text-align: left"><strong>Group Presentations and Discussion</strong></td>
    </tr>
  </tbody>
</table>

<p><br /></p>

<h3 id="instructors">Instructors:</h3>
<p>Ryan Mills (RM) <br />
Alan Boyle (APB)
Hui Jiang (HJ)<br />
<br /></p>

<h3 id="class-questionnaire">Class Questionnaire</h3>
<p>Please help us improve this course by completing this <a href="http://tinyurl.com/bioboot-2016">questionnaire</a>.</p>

<p><br /></p>

<h3 id="data-sets">Data Sets:</h3>
<ul>
  <li>Genotype data for 465 individuals
    <ul>
      <li><a href="ftp://ftp.ebi.ac.uk/pub/databases/microarray/data/experiment/GEUV/E-GEUV-1/genotypes/">Remote site</a></li>
      <li>Local FLUX directory: /scratch/biobootcamp_fluxod/remills/bioboot/geuvadis/genotypes</li>
    </ul>
  </li>
  <li>Expression data for 465 individuals
    <ul>
      <li><a href="ftp://ftp.ebi.ac.uk/pub/databases/microarray/data/experiment/GEUV/E-GEUV-1/analysis_results/">Remote site</a></li>
      <li>Local FLUX directory: /scratch/biobootcamp_fluxod/remills/bioboot/geuvadis/analysis_results</li>
    </ul>
  </li>
</ul>

<p><br /></p>

<h3 id="project-resources">Project Resources</h3>
<ul>
  <li><a href="https://github.com/bioboot/web-2016/blob/gh-pages/class-material/handout_day5_intro-eqtl.pdf">eQTL Introduction Slides</a></li>
</ul>

<p><br /></p>

<h3 id="analysis-notebooks">Analysis notebooks</h3>
<ul>
  <li><a href="https://github.com/bioboot/web-2017/blob/gh-pages/class-material/Day5.ipynb">ipython notebook - eQTL exercise</a></li>
  <li><a href="https://github.com/bioboot/web-2017/blob/gh-pages/class-material/Day5_solution.ipynb">ipython notebook - eQTL exercise, with solutions</a></li>
</ul>

<p><br /></p>
<ul>
  <li>Using what you’ve learned the previous days, login to FLUX, navigate to your personal /scratch folder, and make a new directory called ‘day5’.</li>
  <li>Download today’s ipython notebook into this directory using wget:
    <ul>
      <li><em>wget https://raw.githubusercontent.com/bioboot/web-2017/gh-pages/class-material/Day5.ipynb</em></li>
    </ul>
  </li>
  <li>
    <p>For this exercise, we will need to run ipython notebook on flux. As with Day 4, we can now make use of an internal University of Michigan tool called <a href="https://connect.arc-ts.umich.edu/">ARC Connect</a> to do of all of this far us. Navigate to this URL and login with your UM account. When prompted, complete the 2-factor authentication. From the ARC Connect screen, choose:</p>

    <ul>
      <li>Select <em>biobootcamp_fluxod</em> under Account</li>
      <li>Select <em>Jupyter Notebook</em> under Sesson type.</li>
      <li>All other values can remain at default</li>
      <li>Press <em>Submit your job</em> and wait for it to be allocated (it might take a few minutes)</li>
    </ul>
  </li>
</ul>


  </article>

</div>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Workshop Instructors</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li><a href="https://sph.umich.edu/faculty-profiles/kang-hyunmin.html">Hyun Min Kang</a></li>
          <li><a href="https://gsbs.uth.edu/faculty/faculty-directory/faculty-profiles.htm?id=88d65672-6183-4972-997c-e7e74598e4ec">Goo Jun</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/htsworkshop">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">htsworkshop</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
