<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>KOGO HTS Workshop 2018</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="/kogo-2018/css/main.css">
  <link rel="canonical" href="http://localhost:4000/kogo-2018/class-material/day3-bam-practice.html">
  <link rel="alternate" type="application/rss+xml" title="KOGO HTS Workshop 2018" href="http://localhost:4000/kogo-2018/feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/kogo-2018/">KOGO HTS Workshop 2018</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        <a class="page-link" href="/kogo-2018/day1/">Day 1</a>
        <a class="page-link" href="/kogo-2018/day2/">Day 2</a>
        <a class="page-link" href="/kogo-2018/setup/">Setup</a>
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <article class="post-content">
    <h3 id="iii-2-accessing-aligned-sequence-reads-in-sambam-format">III-2. Accessing aligned sequence reads in SAM/BAM format</h3>

<p>In this part, we will briefly learn how to access SAM/BAM format files for genomic analysis.</p>

<ol>
  <li>Use <code class="highlighter-rouge">samtools</code> to view a BAM file within a specific region
    <pre>
$ samtools view data/HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam 20:10000000 | less
$ samtools view data/HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam 20:10000000 | less -S </pre>
    <ul>
      <li>What is the difference? Which way is more convenient to you?</li>
      <li>Can you interpret what you see on the screen?</li>
    </ul>
  </li>
  <li>Understand how to read the BAM file header.
    <pre>
$ samtools view -H data/HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam | less -S </pre>
    <ul>
      <li>Which version of reference genome is it aligned to?</li>
      <li>What is the sample ID for the sequence reads? In which center and by which platform was it sequenced?</li>
      <li>Which software tools are used to produce the data?</li>
    </ul>
  </li>
  <li>Examine sequence reads at a potential variant site
    <ul>
      <li>From the 1000 Genomes VCF, see where the individual have protein-altering non-reference genotypes.
        <pre>
$ tabix data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5.20130502.genotypes.annotation.vcf.gz 20:10000000-11000000 | cut -f 1,2,4,5,8,10 | grep -v "0|0" | grep missense
20	10603750	G	A	AC=328;AF=0.0654952;AN=5008;NS=2504;DP=19050;EAS_AF=0.004;AMR_AF=0.1066;AFR_AF=0.0121;EUR_AF=0.1322;SAS_AF=0.1033;AA=G|||;CSQ=A|ENSG00000149346|ENST00000334534|Transcript|missense_variant|1130|950|317|R/Q|cGa/cAa|||1|tolerated(0.38)|benign(0.008)||||;GENCODE=CDS_chr20:10603307-10604024;ERB=A||proximal_78873|Regulatory_Feature|proximal_enhancer	0|1 </pre>
      </li>
      <li>Use <code class="highlighter-rouge">samtools tview</code> to visualize the sequence reads in the position
        <pre>
$ samtools tview data/HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam data/hs37d5.fa </pre>
      </li>
      <li>Press <code class="highlighter-rouge">g</code> to enter the position to navigate to</li>
      <li>Type <code class="highlighter-rouge">20:10603750</code> to jump to the variant site</li>
      <li>Does it look like a variant site or not?</li>
      <li>Press <code class="highlighter-rouge">?</code> for more options.</li>
      <li><a href="http://www.broadinstitute.org/igv">IGV</a> is more comprehensive tools to visualize BAM and sequence data.</li>
      <li><a href="http://genome.ucsc.edu">UCSC Genome Browser</a> also provides an interface to load aligned sequence reads.</li>
    </ul>
  </li>
  <li>Create a pileup of sequence reads to process the sequence data by genomic coordinate
    <ul>
      <li>Generate the pileup of sequence reads around the variant site examined above
        <pre>
$ samtools mpileup -r 20:10603745-10603755 -f data/hs37d5.fa data/HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
20	10603745	T	170	,$,...,.....,,,.,,,,.,.,.,,.,,,,...,,,,,.....,.,,,,,,,,,,,,,,,.,,..,,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,^],	3LBBHOACGEGPPJBPOPNHPIPDPNFQ@PPG@GPNLPPE@EEAPiPPHQPPJOPPPJQQLAQKEHQGKQLJLPGDHQQQQGQQLKIPLDGBKKMMFBIQIJQEPJG@NJDQFKQQDIMKEFJGBDQOJEKMOFMGJKHHDMHJJLNKJNGGMNJNMIFMB&lt;KJEIJKFE
20	10603746	G	174	,...,.....,,,.,,,,.,.,.,,.,,,,...,,,,,.....,.,,,,,,,,,,,,,,,.,,..,,$,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,^].^].^].^],^],	LFFHPHHHFFPKPHQPQNJKNQGQNHQ;QQHFIQNQQQJGDHFQlQQFQQQQPQQKQQQQGQQOSQ4QQLQLQIFQKIQLRQRSQLQRRRHSRMRRHLQRRQKQQ&lt;DMNQRLRFRJRKRK?JQRJQRMMFLMRKDKFQIIMICNELFQR?ILQQQIKDK2GNMCFHNCC323CC
20	10603747	A	175	,...,.....,,,.,,,,.,.,.,,.,,,,...,,,,,.....,.,,,$,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,^],^],	JIGHNHGHFHCOOEONOLGOJO?PLFOEOOGDGJ?OOOHHHJGOlPK4POOPNPOOOPOOCOOMNOOOKPKONFODOKPMPPPPPOPELEMPLQLGLPPPJDOP:CLHKPLLLPIIKPJADKGJKOIJHHCKLELHKLILIBLFLHOOGDINKNLKGK?OFJCFHLEEL:&gt;FDBB
20	10603748	T	174	,$...,.....,,,.,,,,.,.,.,,.,,,,...,,,,,.....,.,,,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,,,	3DADNFEG@FKHN0OHJLHOHICOLCO?JKD&lt;BO;OOOHDEIBJqFJPOJONOOFOPFE?OILFOOOLPLKL?OOOKPJPLLOOKPCLCPELMKEMOPPLIJO@:MMPPKKPPIMLPIDIOPIOOLLJGPPECLHOIHLEKIJJJKOHILOOOMBFLBOMJEJ@MGIL7&gt;IFED
20	10603749	C	172	...,.....,,,.,,,,.,.,.,,.,,,,...,,,,.....,.,,,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,,$.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,,,	;DEGCCD@E@DL=QGH&lt;IRJQ=R&gt;EI=MQI0DHIHQI?DIBJnRLRRQFQIHQHIBQ=QRHHIIREIDIRARLRL&gt;MIERRIQRDMALRMNLGJBIRRCMI2BFNIRLQMJIEFQLB&gt;HQH?Q@FAMQIL@DBQGDJHKNFHGBK@:KKQPMK8E?KOJ:BDO&gt;FN5&gt;?@ED
20	10603750	G	174	.AA,A..AA,,aAa,,,$Aa.a.aA,,aaA..,,,aA.AA.aAaaaaaaa,,a,,,aAaaA.,,a,,,,..aaaa,.,,Aa,aa..AAAaa.A,,,AaAa,.,a,a.Aa,.,,AAa,,AA,A,,,a..a,,,AaAaaaaaaaA.a,a..AaA,,a.AA,,,A,a.AA,,aaa^].^],^],	@8INIA@=JPQNBONL4HOEO&gt;OGQ?OOF@@QQQOJ?GK=OsPOPOOPNQPNPQQO?OPI=QQPLQKQJ;OOPPQFQQQOQOP@IGQQKPIKRQJQOKOQ<RLMPIQBRBMMMMADPQFQPKFKLIIGFILPGJLPLMFHHPIGJMHIPMLEL>HOPIIMO&lt;EB@&gt;EFED?3CE
20	10603751	A	174	..,.....,,,.,,,.,.,.,.,,,,...,,,,.....,.,,,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,,,,..,,^],	8FLEIIEDNONFOLJKOJOJOCO0OOGCEOPOOGHEDFOmPOPOOPNOONOPPO@OOLHOOOKPJOGEOOOOPCPPMOKOPEHAMQLPKKPPPMPKIODJQPPJLPPIPQPLDFHLBOPLLIMOLIHKIPIJMIILHMGPOEFKOOIOKDL8OOPHILOBFD;&gt;GFED;/LGGE
20	10603752	T	174	.$.,.....,,,.,,,.,.,.,.,,,,...,,,,.....,.,,,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,,,,..,,,	YDLFgDAFOPDFPGEGPIHGPJO9PPHA@PPOOIGGG@PqPPPPOPNPKIHQLO0PPLFLPPLQKPHDOPKJPLPPPJPPJEJEPPLQLJLLQPQCPPDQMMMIPLQEMLPKCAOPIPPJLLMPPEFIMOIHKIEKHIICOCGLOOOKJEMDONJGHMN0DNDCDI@HALLIHH
20	10603753	T	176	.,.....,,,.,,,$.,.,.,.,,,,...,,,,.....,.,,,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,,,,..,,,,^].^].	EKDcFEEMNMAID3GNHNGNJJ9NNG;ANNNNJHHJENkOOONNOLNIMMOOF?NJKENOOIOJNICNJOOOGOOPGONOHKFKLJOKKKOKKOJDNBOKPIKPOPPOKPB@FNP@OPILEJPPHFJFOBJCFHGDBCOP:JDOOOGJBJ=OHODEGN4BN;OEFBE@LMDF&gt;D33
20	10603754	A	176	.$,$.....,,,.,,$.,.,.,.,,,...,,,,.....,.,,,,,,,,,,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,,...,,,,,..,,,,..^].^],	C3DkDD?PQPdR3GRHRERHMQQB=ERRMQEEEDDQrRRRRQMNRMLJRMQ@RRJFRRRMRMQFDQRRRRLRRJRLRSFKFKPNSKDRRRKRARRASSSSJORSJSSJD@NQJGRONNGOJILIOHOHHRLNOIKIOOINONOHNIHOANNOKIJNDIIANJLGKJHLJKIJKK3E
20	10603755	G	172	.....,,,.,.,.,.,.,,,...,,,,.....,.,,,,,,,,,$,,,,,.,,..,,,,,,,..,,,,,.,,.,,,,.....,,..,,,.,.,,.,,,,..,,.,,..,,,..,.,,,,..,,,,.,.,,,,,,,..,,,...,.,,,...,,,.,...,,,,,..,,,,...,	GjFFGNONgPJPHPGPMKPPIFEPPPOIGGJEPmPLQPPQMP3OOQPODPPMGPPPLQIOREKQPKQNLLIQJPQHMHMRLQM?QQQRQJPPGRQQMLRLRRRKRGC@OQJQJKAHMRRHHQKQB?HILNCICQI?LNLQLDL?I7QPQCGGPGPBKDG9HAOPKHEDNNLC  &lt;/pre&gt;
</RLMPIQBRBMMMMADPQFQPKFKLIIGFILPGJLPLMFHHPIGJMHIPMLEL></mpileup></pre>
        <ul>
          <li>The first four columns represent chromosome, position, reference base, and depth.</li>
          <li>The fifth column represent the base
            <ul>
              <li><code class="highlighter-rouge">,</code> or <code class="highlighter-rouge">.</code> represents matching bases to reference genome.</li>
              <li><code class="highlighter-rouge">^]</code> or <code class="highlighter-rouge">$</code> represents beginning or end of the reads</li>
              <li>Upper/lowercase alphebets represents mismatch to the refernece genome.</li>
            </ul>
          </li>
          <li>Which position appears to have a variant?</li>
          <li>The pileup format from samtools is explained <a href="http://samtools.sourceforge.net/pileup.shtml">HERE</a> in detail.</li>
        </ul>
      </li>
    </ul>
  </li>
</ol>

<p>Congratulations! Well done in understanding the basics of SAM/BAM format.
<br />
Now let’s go to next step <a href="../class-material/day3-gtf-practice.html">Representing genes and transcripts using GTF and genePred format</a>
, or go back to <a href="../day3">Day 3 Overview</a>.</p>

  </article>

</div>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Workshop Instructors</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li><a href="https://sph.umich.edu/faculty-profiles/kang-hyunmin.html">Hyun Min Kang</a></li>
          <li><a href="https://gsbs.uth.edu/faculty/faculty-directory/faculty-profiles.htm?id=88d65672-6183-4972-997c-e7e74598e4ec">Goo Jun</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/htsworkshop">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">htsworkshop</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
