<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>KOGO HTS Workshop 2018</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="/kogo-2018/css/main.css">
  <link rel="canonical" href="http://localhost:4000/kogo-2018/class-material/day3-gtf-practice.html">
  <link rel="alternate" type="application/rss+xml" title="KOGO HTS Workshop 2018" href="http://localhost:4000/kogo-2018/feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/kogo-2018/">KOGO HTS Workshop 2018</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        <a class="page-link" href="/kogo-2018/day1/">Day 1</a>
        <a class="page-link" href="/kogo-2018/day2/">Day 2</a>
        <a class="page-link" href="/kogo-2018/setup/">Setup</a>
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <article class="post-content">
    <h3 id="iii-3-representing-genes-and-transcripts-using-gtf-and-genepred-format">III-3. Representing genes and transcripts using GTF and genePred format</h3>

<p>In this part, we will learn how to GENCODE GTF and UCSC genePred format for representing genes and transcripts.</p>

<ol>
  <li>UCSC Genome Bioinformatics sites provides useful resources for understanding human genomes. For protein-coding genes, it provides a ‘refFlat’ file containing the coordinates or transcripts and coding sequences in a succint way.
    <ul>
      <li>Download the refFlat file from UCSC FTP site
        <pre>
$ wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/refFlat.txt.gz </pre>
      </li>
      <li>Have a peek of the file contents
        <pre>
$ zless refFlat.txt.gz
DRD4    NM_000797       chr11   +       637304  640705  637304  640603  4       637304,639432,639647,640400,    637589,639545,640306,640705,
GABRA2  NM_000807       chr4    -       46251580        46392056        46252324        46390723        10      46251580,46263942,46305476,46307584,46312189,46314512,46334631,46388090,46390652,46391751,      46252621,46264145,46305629,46307728,46312272,46314733,46334699,46388206,46390733,46392056,
BHLHE23 NM_080606       chr20   -       61637330        61638387        61637400        61638126        1       61637330,       61638387,
POR     NM_000941       chr7    +       75544419        75616173        75583310        75615799        16      75544419,75583306,75601730,75608768,75609656,75610365,75610834,75611541,75612837,75613055,75614094,75614375,75614896,75615240,75615476,75615654,        75544497,75583498,75601779,75608897,75609806,75610490,75610924,75611640,75612954,75613174,75614276,75614525,75615167,75615386,75615559,75616173,
... </pre>
      </li>
      <li>What does each column mean? See <a href="http://hgdownload-test.cse.ucsc.edu/goldenPath/ce3/database/refFlat.sql">HERE</a> if you understand what the SQL table is. Otherwise, see below.
        <ol>
          <li><em>Canonical</em> gene name (or gene symbol). This is the most widely used representation of gene names in human-readable form.</li>
          <li>mRNA transcript ID from NCBI. This is less ambiguous than canonical gene name, when multiple transcript encodes the same gene, but harder to be understood by human.</li>
          <li>Chromosome name where the gene is located on.</li>
          <li>Strand of transcriptional direction.</li>
          <li>The leftmost base position of gene for all exons.</li>
          <li>The rightmost base position of gene for all exons.</li>
          <li>The leftmost base position of coding sequences (excluding UTRs).</li>
          <li>The rightmost base position of coding sequences (excluding UTRs).</li>
          <li>The list of start position for each coding sequence.</li>
          <li>The list of end position for each coding sequence.</li>
          <li>The number of exons.</li>
          <li>The list of start position for each exon.</li>
          <li>The list of end position for each exon.</li>
        </ol>
      </li>
    </ul>
  </li>
  <li><strong>(DIY)</strong> Now let’s try to answer to simple questions about the genes using the refFlat database.
    <ul>
      <li><strong>How many total refFlat entries exist? What is the unique number of canonical gene names and transcript IDs that exist in the refFlat database?</strong>
        <ul>
          <li>The answers are <code class="highlighter-rouge">54196</code>, <code class="highlighter-rouge">26397</code>, and <code class="highlighter-rouge">50354</code>, respectively.</li>
          <li>You can obtain these answers by combining <code class="highlighter-rouge">zcat</code>, <code class="highlighter-rouge">cut</code>, <code class="highlighter-rouge">sort</code>, <code class="highlighter-rouge">uniq</code>, and <code class="highlighter-rouge">wc</code> commands.</li>
          <li>If you need more help, <a href="../class-material/day3-answers.html#a-how-many-total-refflat-entries-exist-what-is-the-unique-number-of-canonical-gene-names-and-transcript-ids-that-exist-in-the-refflat-database">CLICK HERE</a></li>
        </ul>
      </li>
      <li><strong>How many different transcript exist for <em>LDLR</em> gene? How are they similar and different?</strong>
        <ul>
          <li>The answer is <code class="highlighter-rouge">5</code></li>
          <li>You can obtain these answers by combining <code class="highlighter-rouge">zcat</code>, <code class="highlighter-rouge">grep</code>, and <code class="highlighter-rouge">wc</code> commands. <code class="highlighter-rouge">-w</code> option is needed for <code class="highlighter-rouge">grep</code> to obtain the correct answer.</li>
          <li>If you need more help, <a href="../class-material/day3-answers.html#b-how-many-different-transcript-exist-for-ldlr-gene-how-are-they-similar-and-different">CLICK HERE</a></li>
        </ul>
      </li>
      <li><strong>For multiple transcripts with the same canonical gene names, are they always located in the same chromosome or not?</strong>
        <ul>
          <li>This could be a little more tricky.</li>
          <li>You can obtain the answers by combining <code class="highlighter-rouge">zcat</code>, <code class="highlighter-rouge">cut</code>, <code class="highlighter-rouge">sort</code>, <code class="highlighter-rouge">uniq</code>, and <code class="highlighter-rouge">wc</code> commands.</li>
          <li>If you need help or wants to check the answer, <a href="../class-material/day3-answers.html#c-for-multiple-transcripts-with-the-same-canonical-gene-names-are-they-always-located-in-the-same-chromosome-or-not">CLICK HERE</a></li>
        </ul>
      </li>
    </ul>
  </li>
  <li><a href="http://www.gencodegenes.org/">GENCODE</a> is a part of ENCODE project to comprehensively identify all genes and their features. While UCSC refFlat database provides information of well known genes, GENCODE resources are more up-to-date to encompass recently identified genes and transcripts. GENCODE uses the Gene Transfer Format (GTF) to represent their contents.
    <ul>
      <li>Have a peek of downloaded contents of GTF file (GENCODE v19)
        <pre>
$ zless data/gencode.v19.annotation.gtf.gz
##description: evidence-based annotation of the human genome (GRCh37), version 19 (Ensembl 74)
##provider: GENCODE
##contact: gencode@sanger.ac.uk
##format: gtf
##date: 2013-12-05
chr1    HAVANA  gene    11869   14412   .       +       .       gene_id "ENSG00000223972.4"; transcript_id "ENSG00000223972.4"; gene_type "pseudogene"; gene_status "KNOWN"; gene_name "DDX11L1"; transcript_type "pseudogene"; transcript_status "KNOWN"; transcript_name "DDX11L1"; level 2; havana_gene "OTTHUMG00000000961.2";
chr1    HAVANA  transcript      11869   14409   .       +       .       gene_id "ENSG00000223972.4"; transcript_id "ENST00000456328.2"; gene_type "pseudogene"; gene_status "KNOWN"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_status "KNOWN"; transcript_name "DDX11L1-002"; level 2; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";
chr1    HAVANA  exon    11869   12227   .       +       .       gene_id "ENSG00000223972.4"; transcript_id "ENST00000456328.2"; gene_type "pseudogene"; gene_status "KNOWN"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_status "KNOWN"; transcript_name "DDX11L1-002"; exon_number 1;  exon_id "ENSE00002234944.1";  level 2; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";     
... </pre>
      </li>
      <li>If you want to understand what each column represent <a href="http://useast.ensembl.org/info/website/upload/gff.html?redirect=no">CLICK HERE</a></li>
      <li>What is your impression to GTF format? What are the advantages and disadvantages of GTF format compared to refFlat format?</li>
    </ul>
  </li>
  <li><strong>(DIY)</strong> Now let’s try to answer to simple questions about the GENCODE databases
    <ul>
      <li><strong>How many genes, transcripts, exons, and CDS exists?</strong>
        <ul>
          <li>The answers are <code class="highlighter-rouge">57,820</code>, <code class="highlighter-rouge">196,520</code>, <code class="highlighter-rouge">1,196,293</code> and <code class="highlighter-rouge">723,784</code>, respectively.</li>
          <li>You can obtain these answers by combining <code class="highlighter-rouge">zcat</code>, <code class="highlighter-rouge">grep</code>, <code class="highlighter-rouge">cut</code>, <code class="highlighter-rouge">sort</code>, <code class="highlighter-rouge">uniq</code>  commands.</li>
          <li>If you need more help, <a href="../class-material/day3-answers.html#a-how-many-genes-transcripts-exons-and-cds-exists-in-gencode-v19-gtf">CLICK HERE</a></li>
        </ul>
      </li>
      <li><strong>See more comprehensive summary statistics <a href="http://www.gencodegenes.org/stats/archive.html#a19">HERE</a>. Do you think you can reproduce these results?</strong></li>
    </ul>
  </li>
</ol>

<p>Congratulations! Well done in understanding GTF and genePred (or refFlat) formats.
<br />
Now let’s go to next step
<a href="../class-material/day3-bed-practice.html">Working with BED files</a>
, or go back to <a href="../day3">Day 3 Overview</a>.</p>

  </article>

</div>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Workshop Instructors</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li><a href="https://sph.umich.edu/faculty-profiles/kang-hyunmin.html">Hyun Min Kang</a></li>
          <li><a href="https://gsbs.uth.edu/faculty/faculty-directory/faculty-profiles.htm?id=88d65672-6183-4972-997c-e7e74598e4ec">Goo Jun</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/htsworkshop">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">htsworkshop</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
