<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>KOGO HTS Workshop 2018</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="/kogo-2018/css/main.css">
  <link rel="canonical" href="http://localhost:4000/kogo-2018/class-material/day3-fasta-practice.html">
  <link rel="alternate" type="application/rss+xml" title="KOGO HTS Workshop 2018" href="http://localhost:4000/kogo-2018/feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/kogo-2018/">KOGO HTS Workshop 2018</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        <a class="page-link" href="/kogo-2018/day1/">Day 1</a>
        <a class="page-link" href="/kogo-2018/day2/">Day 2</a>
        <a class="page-link" href="/kogo-2018/setup/">Setup</a>
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <article class="post-content">
    <h3 id="iii-1-understanding-how-to-access-fasta-files">III-1. Understanding how to access FASTA files</h3>

<p>In this part, we will learn how to access FASTA files for genomic analysis.</p>

<ol>
  <li>If you’re logged out, login back to the server.
    <pre>
$ ssh [your_id]@flux-login.engin.umich.edu
(...Enter login credentials...)
$ cd bioboot/day3
$ export PATH=/home/hmkang/bioboot/bin:${PATH} </pre>
  </li>
  <li>
    <p>Explore the 1000 GRCh37(+decoy) reference in the data directory.</p>

    <pre> $ less data/hs37d5.fa </pre>

    <blockquote>
      <p>This version of FASTA file is the most widely used version of FASTA file in DNA sequence mapping, including in the 1000 Genomes Project.</p>
    </blockquote>
  </li>
  <li><strong>(DIY)</strong> List all chromosome names available in the FASTA file.
    <ul>
      <li>Hint 1 : use <code class="highlighter-rouge">grep</code> command.</li>
      <li>Hint 2 : Adding <code class="highlighter-rouge">^</code> in the beginning of the <code class="highlighter-rouge">grep</code> query searches for the string at the beginning of the line.</li>
      <li>Solution : You can list all chromosomes using the following command (make sure to include quotation marks)
        <pre> $ grep "^&gt;" data/hs37d5.fa | less </pre>
      </li>
      <li>Is this an efficient way to get the answer? Why?</li>
    </ul>
  </li>
  <li><strong>(DIY)</strong> Alternative method - use FASTA index files produced by <code class="highlighter-rouge">samtools</code>.
    <ul>
      <li><code class="highlighter-rouge">samtools</code> provides a convenient way to explore FASTA files.</li>
      <li>Running <code class="highlighter-rouge">samtools faidx</code> on a FASTA file will create <code class="highlighter-rouge">.fai</code> file (you need a write permission to the directory).
        <pre> $ samtools faidx [input.fasta] </pre>
      </li>
      <li>FASTA index file is human-readable.
        <pre>
$ less data/hs37d5.fa.fai
1       249250621       52      60      61
2       243199373       253404903       60      61
3       198022430       500657651       60      61
4       191154276       701980507       60      61
5       180915260       896320740       60      61
6       171115067       1080251307      60      61
7       159138663       1254218344      60      61
8       146364022       1416009371      60      61
9       141213431       1564812846      60      61
10      135534747       1708379889      60      61
11      135006516       1846173603      60      61
... </pre>
      </li>
      <li>What does each column represent?
        <ol>
          <li>Chromosome (or sequence) name</li>
          <li>Number of nucleotides (or characters) in each chromosome (or sequence).</li>
          <li>The byte offset where each chromosome sequence starts
            <ul>
              <li>For example, you can access the beginning of chromosome 2 this way.</li>
            </ul>
            <pre>
$ tail -c +253404904 data/hs37d5.fa | head
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN </pre>
          </li>
          <li>Number of bases pairs appearing each line.</li>
          <li>Number of bytes used for each line.</li>
        </ol>
      </li>
    </ul>
  </li>
  <li>Use <code class="highlighter-rouge">samtools faidx</code> to access an arbitrary region of genome.
    <ul>
      <li>If <code class="highlighter-rouge">.fai</code> file is available, you can randomly access the FASTA file, without scanning from the beginning of the file</li>
      <li>For example, to look at the 1000 bases starting from the 10M-th bases in chromosome 20,
        <pre>
$ samtools faidx data/hs37d5.fa 20:10000000-10001000
&gt;20:10000000-10001000
TTGTTTACTACAGCTTTGTAGTAAATTTTGAACTCTAAAGTGTTAGTTCTCTAACTTTGT
TTGTTTTTCAAGAGTGTTTTGACTCTTCTTACTGCATCCCTTGCATTTCCATATGGACTT
TATAATCAGCCATGTCAACTTCTGCAAGAAAGACAGCTAGGATTTTGATAAGGATTGTGT
TGAATCTGTAGTCCAATTTTGGGAATACTACCATGTTAACATCGTCTTCCCATCCATGCA
CGTGCAATAGCTTACCATTTATTTGGTCTTCCTCAATTTCTGTCAACAATGATTTGTAGT
TTTCAGTTGCAAGTCTTGCACTTCTTTTGTTAAACTTTTTCCAAATATTTATTCTTCTTA
ATTACATAATTCTCATAATTAATAAAATTCTGAAATTTTCTTAATTTCATTTTTGTGGCC
ATGCACTTTAAAACTTGCCTTTTAACAAGACTTCCAGATGATTCTTGTGAACATTAAGTT
TGTGAAGTGCTTCTCTATTGACAAACTGAACTCACGTGACATTCCACACAGCACTGGCAA
ATTCTGTCCCGTAACTCCGCTAGCTCTCCAACAAGAAGTGACTTGACGCAGCCCAAGGTT
ACTACTTAACACAATGAATTAAATGTTTTTAAATAAGAGGAAGCAATAAGATTCTAAAGG
CTTTTCTGTTTTAATTTTCATGCAATGGAAAACTGGTATTAAATATCTATTTAATTAGGA
GGAAACTACAATGCTGACTTTTGTCTGAATTATGTAGATAAGTGATTCATTTGAAACAAT
TATTTTGATAATTGTCAATTATCCATTTCATTTTAATGCATTTTTTATTCTTTTTTCAAA
AATAGCAACAATTACAACAGTTAAACCTTATAATGAATATGTTTCCTAAACCCTGTTCTA
CTTTCTGGTTCCAGATCTGACACCAATTACCTTTCTGATTTTGGACAAACCACTTAATAT
TTGTAACTTACAATTACTTCAACTGAATAATAAAAGAATTG </pre>
      </li>
    </ul>
  </li>
  <li>The FASTA file used above is GRCh37 sequence used for the 1000 Genomes project. Another widely used version of GRCh37 sequence can be downloaded from UCSC (hg19).
    <ul>
      <li>The FASTA and index file is available.
        <pre> 
$ less data/ucsc.hg19.fasta.fai
chrM    16571   6       50      51
chr1    249250621       16915   50      51
chr2    243199373       254252555       50      51
chr3    198022430       502315922       50      51
chr4    191154276       704298807       50      51
chr5    180915260       899276175       50      51
chr6    171115067       1083809747      50      51
chr7    159138663       1258347122      50      51
chr8    146364022       1420668565      50      51
... </pre>
      </li>
      <li>What are similarities and differences between the two versions of FASTA files?</li>
    </ul>
  </li>
  <li>Note the difference between the representations of chromosome names.
    <ul>
      <li>
        <p>Running the same command on the new FASTA file would not work</p>

        <pre>
$ samtools faidx data/ucsc.hg19.fasta 20:10000000-10001000 </pre>
      </li>
      <li>
        <p>Adding ‘chr’ in the chromosome name will give the similar results, except that repeated sequences are represented in lowercases.</p>

        <pre>
$ samtools faidx data/ucsc.hg19.fasta chr20:10000000-10001000   
&gt;chr20:10000000-10001000
Ttgtttactacagctttgtagtaaattttgaactctaaagtgttagttctctaactttgt
ttgtttttcaagagtgttttgactcttcttactgcatcccttgcatttccatatggactt
tataatcagccatgtcaacttctgcaagaaagacagctaggattttgataaggattgtgt
tgaatctgtagtccaattttgggaatactaccatgttaacatcgtcttcccatccatgca
cgtgcaatagcttaccatttatttggtcttcctcaatttctgtcaacaatgatttgtagt
tttcagttgcaagtcttgcacttcttttgttaaactttttccaaatatttattcttctta
attacataattctcataattaataaaattctgaaattttcttaatttcatttttgTGGCC
ATGCACTttaaaacttgccttttaacaagacttccagatgattcttgtgaacattaagtt
tgtgaagtgctTCTCTATTGACAAACTGAACTCACGTGACATTCCACACAGCACTGGCAA
ATTCTGTCCCGTAACTCCGCTAGCTCTCCAACAAGAAGTGACTTGACGCAGCCCAAGGTT
ACTACTTAACACAATGAATTAAATGTTTTTAAATAAGAGGAAGCAATAAGATTCTAAAGG
CTTTTCTGTTTTAATTTTCATGCAATGGAAAACTGGTATTAAATATCTATTTAATTAGGA
GGAAACTACAATGCTGACTTTTGTCTGAATTATGTAGATAAGTGATTCATTTGAAACAAT
TATTTTGATAATTGTCAATTATCCATTTCATTTTAATGCATTTTTTATTCTTTTTTCAAA
AATAGCAACAATTACAACAGTTAAACCTTATAATGAATATGTTTCCTAAACCCTGTTCTA
CTttctggttccagatctgacaccaattacctttctgattttggacaaaccacttaatat
ttgtaacttacaattacttcaactgaataataaaagaattg </pre>
      </li>
    </ul>
  </li>
  <li><strong>(DIY)</strong> Are the two subsets of sequences in the above identical to each other?
    <ul>
      <li>Hint 1 : You can use <code class="highlighter-rouge">diff</code> utility to compare the contents between the two files.</li>
      <li>Hint 2 : If you want to convert a string to uppercase, a simple <code class="highlighter-rouge">awk</code> one-liner can do it.
        <pre> $ (previous-command) | awk '{ print toupper($0); }' | (next-command) </pre>
        <ul>
          <li>Or alternatively, you can use <code class="highlighter-rouge">tr [:lower:] [:upper:]</code>. See <code class="highlighter-rouge">man tr</code> for detailed usage of <code class="highlighter-rouge">tr</code> command.</li>
        </ul>
      </li>
      <li>Need help or want to check solutions? <a href="../class-material/day3-answers.html#are-the-two-subsets-of-sequences-in-the-above-identical-to-each-other">CLICK HERE FOR SOLUTIONS</a></li>
      <li>Each FASTA file may have slight differences. It is important to understand where the FASTA file I am using is coming from (and you should be able to reproduce the FASTA file on your own).</li>
      <li>If two FASTA files are based on the same genome build, for the primary chromosomes, the sequences will be identical (except for the upper-lower cases and handling of ambiguous bases).</li>
    </ul>
  </li>
</ol>

<p>Congratulations! Well done in accessing FASTA files.
<br />
Now let’s go to next step <a href="../class-material/day3-bam-practice.html">Accessing aligned sequence reads in SAM/BAM format</a>, or go
back to <a href="../day3">Day 3 Overview</a>.</p>

  </article>

</div>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Workshop Instructors</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li><a href="https://sph.umich.edu/faculty-profiles/kang-hyunmin.html">Hyun Min Kang</a></li>
          <li><a href="https://gsbs.uth.edu/faculty/faculty-directory/faculty-profiles.htm?id=88d65672-6183-4972-997c-e7e74598e4ec">Goo Jun</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/htsworkshop">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">htsworkshop</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
