<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>KOGO HTS Workshop 2018</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="/kogo-2018/css/main.css">
  <link rel="canonical" href="http://localhost:4000/kogo-2018/class-material/day3-fastq-unix-practice.html">
  <link rel="alternate" type="application/rss+xml" title="KOGO HTS Workshop 2018" href="http://localhost:4000/kogo-2018/feed.xml" />
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/kogo-2018/">KOGO HTS Workshop 2018</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        <a class="page-link" href="/kogo-2018/day1/">Day 1</a>
        <a class="page-link" href="/kogo-2018/day2/">Day 2</a>
        <a class="page-link" href="/kogo-2018/setup/">Setup</a>
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <div class="post">

  <article class="post-content">
    <h3 id="i-1-exploring-fastq-files-in-unix">I-1 Exploring FASTQ files in UNIX.</h3>

<p>In this part, we will learn how to view FASTQ files from UNIX file system.</p>

<ol>
  <li>Login to the server, as you learned from <a href="../day1">Day 1</a>, using
    <ul>
      <li>Terminal (Mac OS X)</li>
      <li>MobaXTerm (Windows)</li>
      <li>or any other SSH client you prefer.</li>
    </ul>

    <pre>
$ ssh [your_id]@flux-login.engin.umich.edu
(...Enter login credentials...)
$ mkdir --p bioboot/day3
$ cd bioboot/day3 </pre>
  </li>
  <li>
    <p>Create a symbolic link to the directory containing the input files, and add <code class="highlighter-rouge">PATH</code></p>

    <pre> $ ln -s /home/hmkang/bioboot/data/ data </pre>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>     - `ln -s [target] [link_name]` creates a *shortcut* of the target file.
     - See `man ln` to see the detailed usage of `ln` 
</code></pre></div>    </div>

    <pre> $ export PATH=/home/hmkang/bioboot/bin:${PATH} </pre>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>     - Verify that the `data` and `bin` directories are correctly configured.
</code></pre></div>    </div>
    <pre>
 $ ls data/
 ls data/
 ALL.chr20.phase3_shapeit2_mvncall_integrated_v5.20130502.genotypes.annotation.vcf.gz      gencode.v19.annotation.gtf.gz                            hs37d5.fa.fai
 ALL.chr20.phase3_shapeit2_mvncall_integrated_v5.20130502.genotypes.annotation.vcf.gz.tbi  HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam      ucsc.hg19.fasta
 bioboot_2015a_R1.fastq.gz                                                                 HG00096.chrom20.ILLUMINA.bwa.GBR.exome.20120522.bam.bai  ucsc.hg19.fasta.fai
 bioboot_2015a_R2.fastq.gz                                                                 hs37d5.fa
    
 $ which samtools
 /home/hmkang/bioboot/bin/samtools </pre>
    <blockquote>
      <p>(Note) You can scroll the box above to the right</p>
    </blockquote>
  </li>
  <li>
    <p>Examine the contents of the FASTQ files</p>

    <pre> $ zless data/bioboot_2015a_R2.fastq.gz </pre>
    <ul>
      <li>
        <p>What do you see? Can you interpret what each line means?</p>
      </li>
      <li>Let’s check whether the two FASTQ files are paired.
        <pre>
$ zcat data/bioboot_2015a_R1.fastq.gz | wc -l
$ zcat data/bioboot_2015a_R2.fastq.gz | wc -l </pre>
      </li>
      <li>Do you think the files are paired? Why?</li>
      <li>Let’s check another way.
        <pre>
$ zcat data/bioboot_2015a_R1.fastq.gz | head
$ zcat data/bioboot_2015a_R2.fastq.gz | head </pre>
      </li>
      <li>What additional information does it tell us?</li>
    </ul>
  </li>
  <li>Display only the sequence read information from FASTQ
    <ul>
      <li>What should we display? Which line numbers?</li>
    </ul>

    <pre> $ zcat data/bioboot_2015a_R2.fastq.gz | awk 'NR % 4 == 2 {print;}' | less </pre>

    <ul>
      <li><code class="highlighter-rouge">awk</code> is a script language useful for quick data manipulation.</li>
      <li>The script above takes each line as input, and print the 2nd line of every four lines (line number NR is 4N+2).</li>
      <li>We won’t introduce <code class="highlighter-rouge">awk</code> today, but will use it in some occasions explaining specific details.
        <ul>
          <li>For more information about <code class="highlighter-rouge">awk</code>, <a href="http://www.linuxfocus.org/English/September1999/article103.html">This link</a> could be a good starting point.</li>
        </ul>
      </li>
      <li><code class="highlighter-rouge">sed</code> and <code class="highlighter-rouge">perl</code> one-liners are also useful tools for manipulating data easily
        <ul>
          <li>…without having to write a separate program!</li>
        </ul>
      </li>
    </ul>
  </li>
  <li><strong>(DIY)</strong> What are the barcodes that appear most frequently?
    <ul>
      <li><strong>Given</strong> : Your input file is <code class="highlighter-rouge">data/bioboot_2015a_R2.fastq.gz</code></li>
      <li><strong>Want</strong> : Use UNIX utilities you learned plus <code class="highlighter-rouge">awk</code> to find out the sequences that most frequently occurs int the file.</li>
      <li>Hint 1 : You need to use UNIX pipe multiple times, like the example below
        <pre> $ zcat data/bioboot_2015a_R2.fastq.gz | awk 'NR % 4 == 2 {print;}' | *** DO SOMETHING HERE *** | less </pre>
      </li>
      <li>Hint 2 : <code class="highlighter-rouge">sort</code> utility sorts input lines alphabetically by default. Adding <code class="highlighter-rouge">-n</code> sorts input lines numerically. Adding <code class="highlighter-rouge">-r</code> sorts input in decreasing order. See <code class="highlighter-rouge">man sort</code> for detailed usage.</li>
      <li>Hint 3 : <code class="highlighter-rouge">uniq</code> utility filter redundant lines appearing consecutively. <code class="highlighter-rouge">uniq -c</code> provides the number of repeats for each input line. See <code class="highlighter-rouge">man uniq</code> for detailed usage</li>
      <li>Need help or done? <a href="../class-material/day3-answers.html#what-are-the-barcodes-that-appear-most-frequently">CLICK HERE FOR THE ANSWER</a></li>
    </ul>
  </li>
  <li><strong>(DIY)</strong> Demultiplex a pair of FASTQ files.
    <ul>
      <li><strong>Given</strong>
        <ul>
          <li>You have a pair of FASTQ files, <code class="highlighter-rouge">data/bioboot_2015a_R1.fastq.gz</code> and <code class="highlighter-rouge">data/bioboot_2015a_R2.fastq.gz</code></li>
          <li>The first file (51bp) contains actual sequence reads.</li>
          <li>The second file (7bp) contains sample barcodes.
            <pre>
Sample1 ACAGTGA
Sample2 CAGATCA
Sample3 GCCAATA
Sample4 TGACCAA
Sample5 TTAGGCA</pre>
          </li>
          <li>Any pair of barcodes differ by 4 or more nucleotides.</li>
        </ul>
      </li>
      <li><strong>Want</strong>
        <ul>
          <li>Write a python program that splits the first FASTQ files into six parts.
            <ul>
              <li>Five for each sample</li>
              <li>and another one for ‘UNKNOWN’ samples.</li>
            </ul>
          </li>
          <li>Find out how many reads belong to each part.</li>
        </ul>
      </li>
      <li>
        <p>Hint 1 : Use <code class="highlighter-rouge">gzip</code> library to read or write compressed files.</p>

        <pre>
import gzip
fi = gzip.open(input_file_name, 'r')  // for reading
fo = gzip.open(output_file_name, 'w') // for writing </pre>
      </li>
      <li>Hint 2 : Simplest way is split based on exact match.
        <ul>
          <li>If the barcode does not match to any of the sample, consider as UNKNOWN.</li>
          <li>This approach may lose may reads if barcodes have sequencing errors.</li>
          <li>Need more hints? <a href="../class-material/day3-answers.html#a-skeleton-code-for-fastq-demultiplexing">CLICK HERE FOR A SKELETON CODE</a></li>
          <li>Need a solution? <a href="../class-material/day3-answers.html#b-answer-for-fastq-demultiplexing---exact-match">CLICK HERE FOR A SOLUTION</a></li>
        </ul>
      </li>
      <li>Hint 3 : Allow up to one mismatch in the barcode
        <ul>
          <li>Because any pair of barcodes are different by 4 or more nucleotides, if there was one base call error in the barcode, it can be unambiguously resolved.</li>
          <li>By modifying <code class="highlighter-rouge">outfs</code> from the exact match code, we can allow up to one mismatch (how?)</li>
          <li>Need a solution? <a href="../class-material/day3-answers.html#c-answer-for-fastq-demultiplexing---allowing-one-mismatch">CLICK HERE FOR A SOLUTION</a></li>
        </ul>
      </li>
      <li>Bored? Try an advanced version : Use the base quality scores to probablistically calculate the best-matching barcode.
        <ul>
          <li>You can use <code class="highlighter-rouge">ord</code> function to convert characters to ASCII code integer.</li>
          <li>Calculate the likelihood of barcode reads for each case to determine the best matching one.</li>
          <li>What would be advantages and disadvantages for this approach?</li>
        </ul>
      </li>
    </ul>
  </li>
</ol>

<p>Congratulations! Well done in exploring FASTQ files in UNIX!
<br />
Now let’s go back to <a href="../day3">Day 3 Overview</a>.</p>

  </article>

</div>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Workshop Instructors</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col  footer-col-1">
        <ul class="contact-list">
          <li><a href="https://sph.umich.edu/faculty-profiles/kang-hyunmin.html">Hyun Min Kang</a></li>
          <li><a href="https://gsbs.uth.edu/faculty/faculty-directory/faculty-profiles.htm?id=88d65672-6183-4972-997c-e7e74598e4ec">Goo Jun</a></li>
        </ul>
      </div>

      <div class="footer-col  footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/htsworkshop">
              <span class="icon  icon--github">
                <svg viewBox="0 0 16 16">
                  <path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/>
                </svg>
              </span>

              <span class="username">htsworkshop</span>
            </a>
          </li>
          

          
        </ul>
      </div>

      <div class="footer-col  footer-col-3">
        <p class="text"></p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
